executor/parallel_apply.go

```diff
@@ -0,0 +1,383 @@
+		e.resultChkCh <- result{nil, err}
+	}
+	e.workerWg.Done()
+}
+
+// fetchAllInners reads all data from the inner table and stores them in a List.
+func (e *ParallelNestedLoopApplyExec) fetchAllInners(ctx context.Context, id int) (err error) {
+	var key []byte
+	for _, col := range e.corCols[id] {
+		*col.Data = e.outerRow[id].GetDatum(col.Index, col.RetType)
+		if e.useCache {
+			if key, err = codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, key, *col.Data); err != nil {
+				return err
+			}
+		}
+	}
+	if e.useCache { // look up the cache
+		atomic.AddInt64(&e.cacheAccessCounter, 1)
+		e.cacheLock.RLock()
+		value, err := e.cache.Get(key)
+		e.cacheLock.RUnlock()
+		if err != nil {
+			return err
+		}
+		if value != nil {
+			e.innerList[id] = value
+			atomic.AddInt64(&e.cacheHitCounter, 1)
+			return nil
+		}
+	}
+
+	err = e.innerExecs[id].Open(ctx)
+	defer terror.Call(e.innerExecs[id].Close)
+	if err != nil {
+		return err
+	}
+
+	if e.useCache {
+		// create a new one in this case since it may be in the cache
+		e.innerList[id] = chunk.NewList(retTypes(e.innerExecs[id]), e.initCap, e.maxChunkSize)
+	} else {
+		e.innerList[id].Reset()
+	}
+
+	innerIter := chunk.NewIterator4Chunk(e.innerChunk[id])
+	for {
+		err := Next(ctx, e.innerExecs[id], e.innerChunk[id])
+		if err != nil {
+			return err
+		}
+		if e.innerChunk[id].NumRows() == 0 {
+			break
+		}
+
+		e.innerSelected[id], err = expression.VectorizedFilter(e.ctx, e.innerFilter[id], innerIter, e.innerSelected[id])
+		if err != nil {
+			return err
+		}
+		for row := innerIter.Begin(); row != innerIter.End(); row = innerIter.Next() {
+			if e.innerSelected[id][row.Idx()] {
+				e.innerList[id].AppendRow(row)
+			}
+		}
+	}
+
+	if e.useCache { // update the cache
+		e.cacheLock.Lock()
+		defer e.cacheLock.Unlock()
+		if _, err := e.cache.Set(key, e.innerList[id]); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *ParallelNestedLoopApplyExec) fetchNextOuterRow(id int, req *chunk.Chunk) (row *chunk.Row, exit bool) {
+	for {
+		select {
+		case outerRow, ok := <-e.outerRowCh:
+			if !ok { // no more data
+				return nil, false
+			}
+			if !outerRow.selected {
+				if e.outer {
+					e.joiners[id].onMissMatch(false, *outerRow.row, req)
+					if req.IsFull() {
+						return nil, false
+					}
+				}
+				continue // try the next outer row
+			}
+			return outerRow.row, false
+		case <-e.exit:
+			return nil, true
+		}
+	}
+}
+
+func (e *ParallelNestedLoopApplyExec) fillInnerChunk(ctx context.Context, id int, req *chunk.Chunk) (err error) {
+	req.Reset()
+	for {
+		if e.innerIter[id] == nil || e.innerIter[id].Current() == e.innerIter[id].End() {
+			if e.outerRow[id] != nil && !e.hasMatch[id] {
+				e.joiners[id].onMissMatch(e.hasNull[id], *e.outerRow[id], req)
+			}
+			var exit bool
+			e.outerRow[id], exit = e.fetchNextOuterRow(id, req)
+			if exit || req.IsFull() || e.outerRow[id] == nil {
```

Comment:
There may be a bug if req.Isfull and return, the next run would enter Line:356.

Reply:
It would not. 
If the `req.IsFull()` is true, the `e.outerRow[id]` must be `nil` then, and in the next round, the check `e.outerRow[id] != nil` will fail so it would not enter the `Line:356`.

Reply:
`the e.outerRow[id] must be nil` where guarentee this ?